components.go

```diff
@@ -0,0 +1,149 @@
+package discordgo
+
+import (
+	"encoding/json"
+)
+
+// ComponentType is type of component.
+type ComponentType uint
+
+// MessageComponent types.
+const (
+	ActionsRowComponent ComponentType = 1
+	ButtonComponent     ComponentType = 2
+)
+
+// MessageComponent is a base interface for all message components.
+type MessageComponent interface {
+	json.Marshaler
+	Type() ComponentType
+}
+
+type unmarshalableMessageComponent struct {
+	MessageComponent
+}
+
+// UnmarshalJSON is a helper function to unmarshal MessageComponent object.
+func (umc *unmarshalableMessageComponent) UnmarshalJSON(src []byte) (err error) {
```

Comment:
This discussion was hashed out [in the slash commands PR](https://github.com/bwmarrin/discordgo/pull/856#discussion_r577693861); especially when we have only one return value, using named returns isn't super great practice. There are quite a few functions like this, and going with explicit returns simply makes them easier to follow, as well as making it harder for subtle bugs to crop up in the future.